operator/controllers/instance_controller.go

```diff
@@ -0,0 +1,461 @@
+/*
+Copyright 2020 SUSE
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"github.com/xeipuuv/gojsonschema"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	servicebrokerv1alpha1 "github.com/SUSE/metabroker/operator/api/v1alpha1"
+)
+
+// InstanceReconciler implements the Reconcile method for the Instance resource.
+type InstanceReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	ProvisioningPodImage string
+}
+
+// +kubebuilder:rbac:groups=servicebroker.metabroker.suse.com,resources=instances,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=servicebroker.metabroker.suse.com,resources=instances/status,verbs=get;update;patch
+
+const instanceReconcileTimeout = time.Second * 10
+
+// Reconcile reconciles an Instance resource.
+func (r *InstanceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), instanceReconcileTimeout)
+	defer cancel()
+
+	log := r.Log.WithValues("instance", req.NamespacedName)
+
+	instance := &servicebrokerv1alpha1.Instance{}
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
```

Comment:
```suggestion
		if errors.IsNotFound(err) {
			// The instance no longer exists; run any deprovisioning steps necessary.
```

This _does_ mean we may end up repeatedly deprovisioning the same instance, right? (Because we may have a instance reconcile request queued up.)

Reply:
I'm not sure I follow your question. I'll accept the code suggestion, though.

Reply:
We can attempt to reconcile multiple times; that means we can end up calling `deprovisioningPod` multiple times. That _sounds_ like it's probably okay, because it's only creating the pod (and associated resources) if they don't exist.   Though I _think_ we could end up trying it after the pod has completely finished.

Reply:
This can only happen if the creation of the resources took too long and it attempts to create again, which causes an error to happen. While not ideal, I can bump the requeue time when creating these resources.

Reply:
In any case, d11f591 should help with any harmless errors.